### PR TITLE
Do not serialize `OP_CODESEPARATOR`

### PR DESCRIPTION
**Pull request by tcharding**
As is done in bitcoin core; when serializing an input `Script` we should not include the `OP_CODESEPARATOR` opcodes.

Closes: #570 

~Only does legacy stuff.~ See comment on issue for full explanation as to why I believe this PR closes the issue.

### Note to reviewers

This is totally new stuff to me. To do this PR I read the issue description and then read the bitcoin core code (primarily the `CTransactionSignatureSerializer` class). I have no other context or knowledge in this area.

Interestingly `FindAndDelete` is _not_ called to remove the `OP_CODESEPARATOR` opcodes when creating the sighash.

Patch 2 adds unit tests that fail if applied before patch 1. It turned out super nice since it adds a bunch of green lines in the current tests for all the test cases that include `OP_CODESEPARATOR`.

Thanks

### CODE REVIEW

File: src/blockdata/script.rs [hunk 1]
```diff
@@ -287,6 +287,16 @@ impl Script {
     /// Creates a new empty script.
     pub fn new() -> Script { Script(vec![].into_boxed_slice()) }
 
+    /// Creates a script identical to `self` but with all instances of `OP_CODESEPARATOR` removed.
+    pub fn rm_op_codeseparator(&self) -> Script {
+        let v = self.0.iter()
+            .filter(|&op| op != &all::OP_CODESEPARATOR.into_u8())
+            .map(|&op| op)
+            .collect::<Vec<u8>>();
```
Comment (apoelstra) [inline src/blockdata/script.rs:295]: I am very surprised that this passes the tests. This deletes every instance of the `CODESEPARATOR` byte, even if those bytes are not actually opcodes (e.g. they may be inside of datapushes).

Comment (tcharding) [reply src/blockdata/script.rs:295]: Correct me if I'm wrong but isn't that what this code does?
```
    /** Serialize the passed scriptCode, skipping OP_CODESEPARATORs */
    template<typename S>
    void SerializeScriptCode(S &s) const {
        CScript::const_iterator it = scriptCode.begin();
        CScript::const_iterator itBegin = it;
        opcodetype opcode;
        unsigned int nCodeSeparators = 0;
        while (scriptCode.GetOp(it, opcode)) {
            if (opcode == OP_CODESEPARATOR)
                nCodeSeparators++;
        }
        ::WriteCompactSize(s, scriptCode.size() - nCodeSeparators);
        it = itBegin;
        while (scriptCode.GetOp(it, opcode)) {
            if (opcode == OP_CODESEPARATOR) {
                s.write(AsBytes(Span{&itBegin[0], size_t(it - itBegin - 1)}));
                itBegin = it;
            }
        }
        if (itBegin != scriptCode.end())
            s.write(AsBytes(Span{&itBegin[0], size_t(it - itBegin)}));
    }
```

Comment (tcharding) [reply src/blockdata/script.rs:295]: ref: bitcoin/src/script/interpreter.cpp

**Comment by tcharding** [comment]
This is wasting precious reviewer time.

ref: https://github.com/rust-bitcoin/rust-bitcoin/issues/777#issuecomment-1073034640

**Comment by sanket1729** [comment]
Seeing that this is already implemented and somewhat easy to review, I don't mind having this. But as stated previously, I don't see too much value in it.

**Comment by tcharding** [comment]
The diff around the added unit tests was kinda sexy :)

File: src/blockdata/script.rs (hunk 1 above)
Comment (apoelstra) [reply src/blockdata/script.rs:295]: No, this code calls `scriptCode.GetOp` which parses the script.

Comment (tcharding) [reply src/blockdata/script.rs:295]: Oh I get you, `GetOp` (`GetScriptOp`) skips the push data but my code does not.

**Pull request closed without merging**